In [23]:
import requests
import lxml.html as lh
import pandas as pd

In [24]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

page = requests.get(url)

doc = lh.fromstring(page.content)

tr_elements = doc.xpath('//tr')

In [25]:
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

This tells us that there are 3 columns per row

In [26]:
tr_elements = doc.xpath('//tr')

col=[]
i=0

for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print ('%d:"%s"'%(i,name))
    col.append((name,[]))

1:"Postcode"
2:"Borough"
3:"Neighbourhood
"


Creating Pandas dataframe  
Every header is appended to a tuple along with empty list.

In [27]:
for j in range(1,len(tr_elements)):
    T=tr_elements[j]
    
    if len(T)!=3:
        break
    
    i=0
    
    
    for t in T.iterchildren():
        data=t.text_content() 
        
        if i>0:
        
            try:
                data=int(data)
            except:
                pass
        
        col[i][1].append(data)
        
        i+=1

In [28]:
[len(C) for (title,C) in col]

[288, 288, 288]

This shows that each of the 3 columns have exactly 288 rows


Creating the pandas data frame


In [29]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

In [30]:
df.head()

Postcode           Borough     Neighbourhood\n
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

Rearranging and renaming columns


In [31]:
df.columns = ['Postcode', 'Borough','Neighbourhood']

cols = df.columns.tolist()
cols

cols = cols[-1:] + cols[:-1]

df = df[cols]

df.head()

Neighbourhood Postcode           Borough
0      Not assigned\n      M1A      Not assigned
1      Not assigned\n      M2A      Not assigned
2         Parkwoods\n      M3A        North York
3  Victoria Village\n      M4A        North York
4      Harbourfront\n      M5A  Downtown Toronto

In [32]:
df = df.replace('\n',' ', regex=True)
df.head()

Neighbourhood Postcode           Borough
0      Not assigned       M1A      Not assigned
1      Not assigned       M2A      Not assigned
2         Parkwoods       M3A        North York
3  Victoria Village       M4A        North York
4      Harbourfront       M5A  Downtown Toronto

Dropping all cells with a borough that's not assigned


In [33]:
df.drop(df.index[df['Borough'] == 'Not assigned'], inplace = True)

df = df.reset_index(drop=True)

df.head(10)

Neighbourhood Postcode           Borough
0         Parkwoods       M3A        North York
1  Victoria Village       M4A        North York
2      Harbourfront       M5A  Downtown Toronto
3       Regent Park       M5A  Downtown Toronto
4  Lawrence Heights       M6A        North York
5    Lawrence Manor       M6A        North York
6      Not assigned       M7A      Queen's Park
7  Islington Avenue       M9A         Etobicoke
8             Rouge       M1B       Scarborough
9           Malvern       M1B       Scarborough

In [40]:
df = df.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.columns = ['Postcode','Borough','Neighbourhood']
df.head(15)

Postcode      Borough                                      Neighbourhood
0       M1B  Scarborough                                     Rouge ,Malvern
1       M1C  Scarborough             Highland Creek ,Rouge Hill ,Port Union
2       M1E  Scarborough                  Guildwood ,Morningside ,West Hill
3       M1G  Scarborough                                             Woburn
4       M1H  Scarborough                                          Cedarbrae
5       M1J  Scarborough                                Scarborough Village
6       M1K  Scarborough        East Birchmount Park ,Ionview ,Kennedy Park
7       M1L  Scarborough                    Clairlea ,Golden Mile ,Oakridge
8       M1M  Scarborough    Cliffcrest ,Cliffside ,Scarborough Village West
9       M1N  Scarborough                        Birch Cliff ,Cliffside West
10      M1P  Scarborough  Dorset Park ,Scarborough Town Centre ,Wexford ...
11      M1R  Scarborough                                  Maryvale ,Wexford
12      M1S  Scarborough                                          Agincourt
13      M1T  Scarborough            Clarks Corners ,Sullivan ,Tam O'Shanter
14      M1V  Scarborough  Agincourt North ,L'Amoreaux East ,Milliken ,St...

In [35]:
df['Neighbourhood'] = df['Neighbourhood'].str.strip()

In [36]:
df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df['Borough']

In [37]:
df[df['Borough'] == 'Queen\'s Park']

Postcode       Borough Neighbourhood
85      M7A  Queen's Park  Queen's Park

In [38]:
df.shape

(103, 3)